prov/tcp/src/tcpx_ep.c

```diff
@@ -0,0 +1,620 @@
+			    uint64_t flags)
+{
+	return -FI_ENOSYS;
+}
+
+static ssize_t tcpx_send(struct fid_ep *ep, const void *buf, size_t len, void *desc,
+			 fi_addr_t dest_addr, void *context)
+{
+	struct fi_msg msg;
+	struct iovec msg_iov;
+
+	msg_iov.iov_base = (void *) buf;
+	msg_iov.iov_len = len;
+	msg.msg_iov = &msg_iov;
+	msg.desc = &desc;
+	msg.iov_count = 1;
+	msg.addr = dest_addr;
+	msg.context = context;
+
+	return tcpx_sendmsg(ep, &msg, 0);
+}
+
+static ssize_t tcpx_sendv(struct fid_ep *ep, const struct iovec *iov, void **desc,
+			  size_t count, fi_addr_t dest_addr, void *context)
+{
+	struct fi_msg msg;
+
+	msg.msg_iov = iov;
+	msg.desc = desc;
+	msg.iov_count = count;
+	msg.addr = dest_addr;
+	msg.context = context;
+	return tcpx_sendmsg(ep, &msg, 0);
+}
+
+
+static ssize_t tcpx_inject(struct fid_ep *ep, const void *buf, size_t len,
+			   fi_addr_t dest_addr)
+{
+	struct fi_msg msg;
+	struct iovec msg_iov;
+
+	msg_iov.iov_base = (void *) buf;
+	msg_iov.iov_len = len;
+	msg.msg_iov = &msg_iov;
+	msg.iov_count = 1;
+	msg.addr = dest_addr;
+
+	return tcpx_sendmsg(ep, &msg, FI_INJECT | TCPX_NO_COMPLETION);
+}
+
+static ssize_t tcpx_senddata(struct fid_ep *ep, const void *buf, size_t len, void *desc,
+			     uint64_t data, fi_addr_t dest_addr, void *context)
+{
+	return -FI_ENOSYS;
+}
+
+static ssize_t tcpx_injectdata(struct fid_ep *ep, const void *buf, size_t len,
+			       uint64_t data, fi_addr_t dest_addr)
+{
+	return -FI_ENOSYS;
+}
+
+
+static struct fi_ops_msg tcpx_msg_ops = {
+	.size = sizeof(struct fi_ops_msg),
+	.recv = tcpx_recv,
+	.recvv = tcpx_recvv,
+	.recvmsg = tcpx_recvmsg,
+	.send = tcpx_send,
+	.sendv = tcpx_sendv,
+	.sendmsg = tcpx_sendmsg,
+	.inject = tcpx_inject,
+	.senddata = tcpx_senddata,
+	.injectdata = tcpx_injectdata,
+};
+
+static void tcpx_set_sockopts(int sock)
+{
+	int optval;
+
+	optval = 1;
+
+	if (setsockopt(sock, SOL_SOCKET, SO_REUSEADDR, &optval, sizeof(optval)))
+		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,"setsockopt reuseaddr failed\n");
+
+	if (setsockopt(sock, IPPROTO_TCP, TCP_NODELAY, &optval, sizeof(optval)))
+		FI_WARN(&tcpx_prov, FI_LOG_EP_CTRL,"setsockopt nodelay failed\n");
+}
+
+static void tcpx_setup_socket(int sock)
+{
+	tcpx_set_sockopts(sock);
+	fi_fd_nonblock(sock);
+}
+
+static int tcpx_ep_connect(struct fid_ep *ep, const void *addr,
+		    const void *param, size_t paramlen)
+{
+	struct tcpx_ep *tcpx_ep = container_of(ep, struct tcpx_ep, util_ep.ep_fid);
+	struct sockaddr *serv_addr = (struct sockaddr *) addr;
+	struct poll_fd_info *fd_info;
+	struct util_fabric *util_fabric;
+	struct tcpx_fabric *tcpx_fabric;
+	size_t addrlen;
+	int ret = FI_SUCCESS;
+
+	util_fabric = tcpx_ep->util_ep.domain->fabric;
+	tcpx_fabric = container_of(util_fabric, struct tcpx_fabric, util_fabric);
+
+	if (!addr || !tcpx_ep->conn_fd)
+		return -FI_EINVAL;
+
+	switch (serv_addr->sa_family) {
+	case AF_INET:
+		addrlen = INET_ADDRSTRLEN;
+		break;
+	case AF_INET6:
+		addrlen = INET6_ADDRSTRLEN;
```

Comment:
addrlen arg to connect( ) should be size of the struct sockaddr_in/sockaddr_in6 rather than size of string representation of the address.